conda/testing/gateways/fixtures.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+
+import socket
+import tempfile
+from pathlib import Path
+
+import pytest
+
+from .helpers import have_minio
+
+
+def s3_server(xprocess):
+    """
+    Mock a local S3 server using `minio`
+
+    This requires:
+    - pytest-xprocess: runs the background process
+    - minio: the executable must be in PATH
+
+    Note it will be given EMPTY! The test function needs
+    to populate it. You can use
+    `conda.testing.helpers.populate_s3_server` for that.
+    """
+    # The 'name' below will be the name of the S3 bucket containing
+    # keys like `noarch/repodata.json`
+    NAME = "s3_server"
+    PORT = 9000
```

Comment:
Let's move the `populate_s3_server` helper function as new `populate_bucket(endpoint, bucket_name, channel_dir)` here as an instance method so we don't have a need for a `helpers.py` file. Then also set a class attribute to refer to the URL of the bucket so we can pass the whole starter instance to the test and have everything encapsulated in one place.

```suggestion
    class Minio:
        name = "minior_s3_server"
        port = 9000

        @property
        def server_url(self):
            return f"http://localhost:{self.port}/{self.name}"

        def __init__(self):
            bucket_dir = tmp_path / self.name
            bucket_dir.mkdir()

        def populate_bucket(self, endpoint, bucket_name, channel_dir):
            ...
```